Scrape constituency names

ondenman opened the pull request:
Scrape member constituency names.

Members are grouped by [district on the wikipedia source page](https://en.wikipedia.org/wiki/List_of_MPs_elected_in_the_Mongolian_legislative_election,_2016). The district name is only displayed in the upper most row of each district (in column[0]). When the district name is present, it is stored. When a row is without a district name (column[0] contains an ordinal number), the stored district name is used.

ondenman committed:
Scrape constituency names

Scrapes the name of constituency for members. Members are displayed grouped by district. The district name is only displayed in the upper most row of each district. When the district name is present, it is stored. When a row is without a district name, the stored district name is used.

chrismytton commented on scraper.rb:
```diff
@@ -9,6 +9,7 @@
 require 'pry'
 
 class Khurai
+
```
This change doesn't seem related to this pull request. I would also add that Rubocop would report this as an offence - "Extra empty line detected at class body beginning". Probably best to amend the commit to remove this change :)

chrismytton commented on scraper.rb:
```diff
@@ -88,9 +86,15 @@ def tr_without_district
+    cell_text = tds[0].text.strip.gsub("\n", ' — ')
+    @current_constituency = cell_text unless cell_text =~ /\d/
+    current_constituency
+  end
```
I don't think you need the `current_constituency` attribute as it's only being used in this method. Perhaps something like this would be better?

``` ruby
  def constituency
    cell_text = tds.first.text.strip.gsub("\n", ' — ')
    cell_text unless cell_text =~ /\d/
  end
```

ondenman replied:
@current_constituency holds the value of the last district found. If the row doesn't have a district cell, @current_constituency is used. So instead of returning nothing, the methods returns @current_constitunecy.

chrismytton replied:
Ah yes, I understand now! In that case I guess you don't need the `attr_accessor :current_constituency` bit? Instead you can just return `@current_constituency` at the end of the method.

ondenman replied:
I've applied the change in a fixup. fc91033

tmtmtmtm replied:
I'm afraid I'm slower than @chrismytton and still don't understand this logic…

tmtmtmtm replied:
Ah, I think I've got it now — it's because you're treating each row as variable data, some of which hold memory from previous ones. Erk. This implementation is quite scary! In general you should think of object state as being immutable. If you find yourself wanting to change it in a loop, then that's a sign that you have another object here trying to escape. Each Membership here should really be a separate object, distinct from the table.

chrismytton commented on scraper.rb:
```diff
@@ -88,9 +86,15 @@ def tr_without_district
       name: 1,
       name__mn: 2,
-      party: 4,
+      party: 4
```
This change also seems unrelated to this pull request (and is inconsistent with `tr_with_district`, which has a trailing comma on the `Hash` literal). I think it's generally best to be consistent with what's in the rest of the file.

chrismytton reviewed (approved):
Woop woop! ✨  🚀  ✨

ondenman committed:
Extract Members and Rows into own classes

ondenman committed:
Rename Rows to Table

ondenman committed:
Extract Row class

Table now creates array of Row objects. Table passes the node for the current row, page url and current constituency to Row. Row implements to_h.

tmtmtmtm committed:
remove attr_reader on members

This is an actual method, not an attribute

tmtmtmtm committed:
remove scraped_page_archive

This is scraping Wikipedia, which has its own public history, so we
don't need to maintain an archive

tmtmtmtm committed:
remove `source`

`source` doesn't really make sense in this context — that's meant to be
a source for the Person data — for a summary page like this it doesn't
really work.

tmtmtmtm committed:
move `page`, `url` etc onto Khurai

These methods belong to the Khurai/Page, not to the Table

tmtmtmtm committed:
instantiate the Page with a URL

We shouldn't hard-code the URL for the page into the class — in theory
this should be reusable against any page that has the same layout (e.g.
for previous or future terms)

tmtmtmtm committed:
remove empty Member class

the Member class isn't adding any value: it's just holding the data it
was given and giving it straight back again.

tmtmtmtm committed:
simplify logic of `constituency`

Setting this back to itself as one strand of a ternary is a little odd.
Let's just set it if the condition holds.

tmtmtmtm committed:
rename `node` as `tds`

It's misleading to call this `node` as that implies it's a single
Nokogiri Element, when actually it's an Array of TD elements.

tmtmtmtm committed:
rename 'cells' to 'cellmap'

Make it more obvious that this is a mapping function

tmtmtmtm committed:
Post-process the constituency

Rather than injecting a constituency into each Row, whether it had one
or not, process the row separately, and then merge the constituency in
afterwards. My working theory is that this will let us simplify the
logic for Row considerably.

tmtmtmtm committed:
shift off the 'constituency' cell where it exists

If we're dealing with a row that _has_ constituency information, remove
it before building the object. This will let us treat every row the
same way.

tmtmtmtm committed:
remove `cellmap_with_district`

We no longer have any rows that have this column, so we can just remove
this method.

tmtmtmtm committed:
remove cellmap_without_district

This is now the only option, so we can inline it into `cellmate`

tmtmtmtm committed:
remove entire cellmap concept

There's no need for this indirection any more – just inline the cell
numbers into the methods that need them.
